### PR TITLE
feat: v2 organizations event types endpoints

### PR DESCRIPTION
**Pull request by supalarry**
## What
Create [organizations-event-types.controller.ts](https://github.com/calcom/cal.com/pull/15775/files#diff-5abbbdf31b2c9d45e486e3f0f7236338e1cbf0f08fa2ac23a43cc525f8bf8fe3) endpoints to create collective, round robin and managed team event types of an organization:

1. Create team event-type `/v2/organizations/:orgId/teams/:teamId/event-types`
2. Get team event-type `/v2/organizations/:orgId/teams/:teamId/event-types/:eventTypeId`
3. Get team event-types `/v2/organizations/:orgId/teams/:teamId/event-types`
4. Get org event-types `/v2/organizations/:orgId/teams/event-types`
5. Update team event-type `/v2/organizations/:orgId/teams/:teamId/event-types/:eventTypeId`
6. Delete team event-type `/v2/organizations/:orgId/teams/:teamId/event-types/:eventTypeId`

## How

1. Create controller for incoming requests - organizations-event-types.controller.ts
2. [organizations-event-types.service.ts](https://github.com/calcom/cal.com/pull/15775/files#diff-c2b15affd1b88772361b63b2023921ad16093a0fdf2893849eb763632c74b25b) then handles creating, reading, updating and deleting team event-types.

It takes incoming request and uses newly created [input.service.ts](https://github.com/calcom/cal.com/pull/15775/files#diff-a6ffcc72cb05971f1c055f35da6670f2901072743410f6c8f5825a9e2ba290f3) to convert request event-type into form compatible with codebase. The input service
re-uses logic of event-types.input.service.ts, but adds team specific logic, like setting metadata and children for managed team events
, setting up hosts for collective and round robin events or handling the assignement of all team members to event type.

hosts for any team event type are submitted in form of `class Host` of create-event-type.input.ts to simplify developer experience,
and in case of managed event-types children event-types are created based on them.

When creating a team event-type, after input is transformed, event-type is first created using trpc create handler and then updated,
because create handler does not accept hosts and other information like assigning all team members. That should be refactored but is out of scope of this PR.

For response [output.service.ts](https://github.com/calcom/cal.com/pull/15775/files#diff-1956f6dfeac5ae7954e98215e945458afb9f2fd983b0958fc0bfb653dc02eb9f) is used to transform database event-types into shape API defines in
`TeamEventTypeOutput_2024_06_14` of [event-type.output.ts](https://github.com/calcom/cal.com/pull/15775/files#diff-fdfffb5ca46b3e58238144c37c879b8a4859bc45fdaa257d7f5fb98f2c1a1ae2). Notably, when creating and updating team event type, the output can be either one object or array of objects. If its a collective or round robin event, then response will be one object. If it is a managed event, then the response is managed team event + managed event that was created for each user specifically, so if we create a managed event type and pass 2 hosts, then response contains an array of 3 objects that were created in database.

3. Org event-type specific database calls are handled in [organizations-event-types.repository.ts](https://github.com/calcom/cal.com/pull/15775/files#diff-80d925bf2513ecbbc3cbb851f2db5d603a7d01dee4497d7f1cbe4a21fb36ae7d)

4. platform-libraries were bumped from 0.0.18 to 0.0.19 to enable system admin
to create team event types for teams that system admin is not part of. See PR #15774 

5. Define `ApiResponseWithoutData` base DTO with status only that is extended by outputs
`v2/src/modules/organizations/controllers/event-types/outputs/teams` so that they don't have to re-define status,
and this class can be re-used in future response DTOs with following structure:
- define data type in platform-types
- define outputs folder in corresponding v2 controller that extends ApiResponseWithoutData and in data has type from platform-types or array of the type.

## Tests
[organizations-event-types.e2e-spec.ts](https://github.com/calcom/cal.com/pull/15775/files#diff-293a7771b8112204ce3086bb48fa224876fce328e8a8dc8eb7227b60d57efb35) tests CRUD operations of collective and managed event-types.

```
npx jest src/modules/organizations/controllers/event-types/organizations-event-types.e2e-spec.ts --config jest-e2e.json
```

## Next steps
Handle individual org user event-types in event-types.controller.ts

### CODE REVIEW

File: apps/api/v2/src/modules/organizations/services/event-types/organizations-event-types.service.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+    await updateEventType({
+      input: { id: eventTypeId, ...bodyTransformed },
+      ctx: {
+        user: eventTypeUser,
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        prisma: this.dbWrite.prisma,
+      },
+    });
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/services/event-types/organizations-event-types.service.ts:121]: is this function making sure that the eventtype belongs to the teamId inside of the bodyTransformed ?

File: apps/api/v2/src/modules/organizations/controllers/event-types/organizations-event-types.e2e-spec.ts [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+      return request(app.getHttpServer())
+        .delete(`/v2/organizations/${org.id}/teams/${team.id}/event-types/${managedEventType.id}`)
+        .expect(200);
+    });
+
+    function evaluateHost(expected: Host, received: Host) {
+      expect(expected.userId).toEqual(received.userId);
+      expect(expected.mandatory).toEqual(received.mandatory);
+      expect(expected.priority).toEqual(received.priority);
+    }
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/event-types/organizations-event-types.e2e-spec.ts:447]: can we add tests for some fail case ? like trying to update an existing event-type that does not belong to the team ?

File: apps/api/v2/src/modules/organizations/controllers/event-types/organizations-event-types.controller.ts [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+@Controller({
+  path: "/v2/organizations/:orgId",
+  version: API_VERSIONS_VALUES,
+})
+@UseGuards(ApiAuthGuard, IsOrgGuard, RolesGuard)
+@DocsTags("Organizations Event Types")
+export class OrganizationsEventTypesController {
+  constructor(private readonly organizationsEventTypesService: OrganizationsEventTypesService) {}
+
+  @Roles("ORG_ADMIN")
```
Comment (ThyMinimalDev) [inline apps/api/v2/src/modules/organizations/controllers/event-types/organizations-event-types.controller.ts:48]: I think we can consider all those endpoints accessible by TEAM_ADMIN

File: apps/api/v2/src/modules/organizations/services/event-types/organizations-event-types.service.ts (hunk 1 above)
Comment (supalarry) [reply apps/api/v2/src/modules/organizations/services/event-types/organizations-event-types.service.ts:121]: Yes - it fetches event type by id, and then if teamId is passed checks if event type team id is same as passed teamId.

File: apps/api/v2/src/modules/organizations/controllers/event-types/organizations-event-types.controller.ts (hunk 3 above)
Comment (supalarry) [reply apps/api/v2/src/modules/organizations/controllers/event-types/organizations-event-types.controller.ts]: Implemented it in @supalarry
[refactor: organisations event-types endpoints accessible by TEAM_ADMIN](https://github.com/calcom/cal.com/pull/15775/commits/6da785a6e525ef32b0e587087573b8098cc2ca2b)

File: apps/api/v2/src/modules/organizations/controllers/event-types/organizations-event-types.e2e-spec.ts (hunk 2 above)
Comment (supalarry) [reply apps/api/v2/src/modules/organizations/controllers/event-types/organizations-event-types.e2e-spec.ts:584]: added in [fix: dont allow creating event-type for non-team member](https://github.com/calcom/cal.com/pull/15775/commits/b371e72fb9206499bf59873c3abe199ed859e48a) and [tests: negative tests](https://github.com/calcom/cal.com/pull/15775/commits/913d43bb5ed551f8b5aae675f07c1b349c68aac0). All tests can be seen in [organizations-event-types.e2e-spec.ts](https://github.com/calcom/cal.com/pull/15775/files#diff-293a7771b8112204ce3086bb48fa224876fce328e8a8dc8eb7227b60d57efb35)

**Review by ThyMinimalDev — APPROVED** [review]
🔥

**Pull request merged**